core/src/solution_submission.rs

```diff
@@ -215,29 +222,60 @@ async fn retry_with_gas_price_increase(
                 solution.clone(),
                 claimed_objective_value,
                 gas_price,
-                if gas_price == gas_cap {
-                    None
-                } else {
-                    Some(BLOCK_TIMEOUT)
-                },
+                Some(BLOCK_TIMEOUT),
             )
             .await;
 
-        let gas_price_can_still_increase =
-            gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR < gas_cap.as_u128() as f64;
+        let before_deadline = deadline
+            .map(|deadline| Instant::now() < deadline)
+            .unwrap_or(true);
+        let gas_price_can_still_increase = gas_price.as_u128() as f64
+            * MIN_GAS_PRICE_INCREASE_FACTOR
+            < gas_price_cap.as_u128() as f64;
+
         // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
+        match result {
```

Comment:
I find this match statement a little hard to follow and it feels a little deeply nested. Maybe it can be refactored a bit? (PS. I have no good ideas on how to do this).

Reply:
Agree. Tried this a bit. I think it ends up changing the whole function. In that case I'd like keep this PR as is and make another one to refactor the function.

Reply:
Now that I think about it, maybe the function can be refactored in this PR already? Seeing as it is already changing significantly to include the `noop_transaction` to cancel the solution submission?

Reply:
For the reviewers and when looking at individual commits later it is cleaner to have this separate since the refactoring will change more things than just adding this functionality.

Reply:
I cannot follow the code in this match statement (I wasn't even able to properly understand a much simpler flavor [here](https://github.com/gnosis/dex-services/pull/1070#discussion_r449536863)). 

Would it be possible to refactor before adding logic? Maybe we can have a method that returns an `enum SolutionSubmissionAction` and change the loop into something more straight-forward such as:

```rust
loop {
  match next_action(previous_gas_price, gas_price_estimating) {
     SolutionSubmissionAction::Resubmit(gas_price) => { ... }
     SolutionSubmissionAction::LeaveInMemPool => break
  }
}
```

and then in this PR change `LeaveInMemPool` to `CancelTx`. `next_action` could have some unit tests describing what action we expect based on the previous gas price, current estimate and time passed.

Reply:
I would also be in favour of a refactor before the change, as this loop has become a bit unruly and would be nice to split up a bit. I like @fleupold's suggestion of a method that tries to submit a solution and returns an enum that can be used to determine what to do next.

Reply:
I'll do the refactor first in a separate PR then rebase this PR.